### DISCUSSION TITLE
RFC: Add `--plan` flag

### DISCUSSION THREAD

**Discussion by jaredpalmer** [vercel/turborepo#380] [RFC]
### Describe the feature you'd like to request

Add support for `--plan` flag that will output JSON that includes information about what is about to happen for a given command. This can be used with scripts such as Vercel's Ignore Build Command and also in Circle and GitLab for dynamic provisioning of CI/CD workflows.

Nx's custom Ignore Command something like this as a reference: https://github.com/nrwl/nx/blob/master/scripts/vercel/ignore-build.sh

### Describe the solution you'd like

Collect up info through a traversal and print out to JSON

- Should this hit caches / wonder if we need an endpoint specifically to check without downloading which we don't yet support? 

### Describe alternatives you've considered

- Possibly need to do another full pass through the graph / topological sort in order to hash
- Split up scheduler so that the graph generation and execution occur in separate steps (so that it can be interrupted if `--plan` is passed)
- Split this into multiple commands like `plz query` does
<img width="400" alt="CleanShot 2022-03-08 at 10 18 41@2x" src="https://user-images.githubusercontent.com/4060187/157267976-00b5930b-9d2b-4b62-bd9a-9e9323e8bd0e.png">

**Comment by wvanrensselaer** [comment]
Wonder if it would make sense to record/cache runtime metrics for plan tasks to optimize future plans. Our CI runners usually have ~20s setup/cleanup time on top of the task being run. So suppose we have 4 uncached tasks to run:

![tasks](https://user-images.githubusercontent.com/35181014/146663533-448d3a86-4797-4d2b-8693-d5688a7bfb5e.png)

Could initially mirror the Turborepo plan in CI using 4 machines as follows:

![pipeline-unoptimized](https://user-images.githubusercontent.com/35181014/146663539-5c0034ef-8947-4c47-9a3a-8b395142a0c0.png)

But in the future we can optimize the CI plan with prior runtime knowledge. In this case it would be better to group B, C, and D on the same machine:

![pipelin-optimized](https://user-images.githubusercontent.com/35181014/146663548-94443aa5-7d5e-458e-acde-f1d5df4b902b.png)

**Comment by felixmosh** [comment]
This task will allow implement a custom github action based on `actions/cache` 🙏🏼

**Comment by cipriancaba** [comment]
Is there any other way atm to retrieve changed packages since a specific commit?

**Comment by bobaaaaa** [comment]
This would be a solution for https://github.com/vercel/turborepo/discussions/800

**Comment by jaredpalmer** [comment]
Now available in `turbo@1.1.7` as `--dry` and `--dry=json`
